scripts/gulpfiles/docs_tasks.js

```diff
@@ -0,0 +1,129 @@
+const { execSync } = require('child_process');
+const { Extractor } = require('markdown-tables-to-json');
+const fs = require('fs');
+const gulp = require('gulp');
+const header = require('gulp-header');
+const replace = require('gulp-replace');
+
+const docsDir = 'docs';
+
+/**
+ * Run API Extractor to generate the intermediate json file.
+ */
+const generateApiJson = function(done) {
+    execSync('api-extractor run --local', { stdio: 'inherit' });
+    done();
+}
+
+/**
+ * API Extractor output spuriously renames some functions. Undo that.
+ * See https://github.com/microsoft/rushstack/issues/2534
+ */
+const removeRenames = function() {
+    return gulp.src('temp/blockly.api.json')
+        .pipe(replace('_2', ''))
+        .pipe(gulp.dest('temp/'));
+}
+
+/**
+ * Run API Documenter to generate the raw docs files.
+ */
+const generateDocs = function(done) {
+    if (!fs.existsSync(docsDir)) {
+        // Create the directory if it doesn't exist.
+        // If it already exists, the contents will be deleted by api-documenter.
+        fs.mkdirSync(docsDir);
+    }
+    execSync(`api-documenter markdown --input-folder temp --output-folder ${docsDir}`, { stdio: 'inherit' });
+    done();
+}
+
+/**
+ * Prepends the project and book metadata that devsite requires.
+ */
+const prependBook = function() {
```

Comment:
Gulp writing intermediate output either doesn't matter or can't be helped here.
The tasks are: 

```js
gulp.series(
    generateApiJson, removeRenames, generateDocs,
    gulp.parallel(prependBook, createToc));
```

`generateApiJson` and `removeRenames` touch the same file (`blockly.api.json`) but that file is generated by `api-extractor` so there is no opportunity to touch that file before it is written.
`generateDocs` and `prependBook` touch the same files (all 3000 of them) but since those files are generated by `api-documenter` again there is no opportunity to touch those files before they are written. api-documenter doesn't have the JS API to turn this into a stream even if I wanted to.
`createToc` creates a new file itself and does not touch the other files, so there is no point in combining it with the others, and is why it's running in parallel with `prependBook`.

The entire process takes 12 seconds and will be something we run once per release. So I don't think it's worth spending any further time optimizing this process until we find it cumbersome in the future.

Reply:
I was thinking that it would be nice to have a task with a single pipeline that does (almost?) everything in one go:
```
function apiStuff() {
  return gulp.src(SRC_DIR)
      .pipe(/* invoke api-extractor on stream; stream now contains a JSON file */)
      .pipe(replace('_2', ''))
      .pipe(gulp.dest(`api/`)  // Writes JSON file to disk to git diff against previous version.
      .pipe(/* invoke api-documentor on stream; stream now contains markdown files */)
      .pipe(header(
          'Project: /blockly/_project.yaml\nBook: /blockly/_book.yaml\n\n'))
      .pipe(/* invoke createToc, which adds an extra file to the stream */)
      .write(`$BUILD_DIR/docs`);  // Writes markdown files to disk.
```

You are right about the incompatibility of api-extractor and api-documentor with Vinyl streams, though TIL of [the gulp-intermediate plugin](https://github.com/robwierzbowski/gulp-intermediate) that is designed to work around that issue.

I concur that it is almost certainly not worth further effort on your part at this time, but I do not promise not to tidy it up myself at some point if the mood strikes.  :-)